src/SNLS_device_forall.h

```diff
@@ -50,10 +56,99 @@ end,                                             \
+   using crview1d = rview1<const double>;
+   using crview2d = rview2<const double>;
+   using crview3d = rview3<const double>;
+
+   using rhost_res = RAJA::resources::Host;
+#if defined(__snls_gpu_active__)
+#if defined(RAJA_ENABLE_CUDA)
+         using rgpu_res = RAJA::resources::Cuda;
+#else //defined(RAJA_ENABLE_HIP)
+         using rgpu_res = RAJA::resources::Hip;
+#endif
+#endif
+
+   using rres = std::variant<
+      rhost_res
+#if defined(__snls_gpu_active__)
+      ,
+      rgpu_res
+#endif
+   >;
+
+   using rrese = RAJA::resources::Event;
+
+   // We really don't care what View class we're using as the sub-view just wraps it up
+   // and then allows us to take a slice/window of the original view
+   // Should probably work out some form of SFINAE to ensure T that we're templated on
+   // is an actual View that we can use.
+   template<class T>
+   class subview {
+   public:
+      // Delete the default constructor as that wouldn't be a valid object
+      __snls_hdev__
+      subview() = delete;
+      // where we don't want any offset within the subview itself
+      __snls_hdev__
+      subview(const int index, T& view) : m_view(view), m_index(index), m_offset(0) {};
+      // sometimes you might want to have an initial offset in your subview when constructing
+      // your subview in which everything appears as 0 afterwards
+      __snls_hdev__
+      subview(const int index, const size_t offset, T& view) : m_view(view), m_index(index), m_offset(offset) {};
+
+      ~subview() = default;
+
+      // Could probably add default copy constructors as well here if need be...
+
+      // Let the compiler figure out the correct return type here as the one from
+      // RAJA at least for regular Views is non-trivial
+      // make the assumption here that we're using row-major memory order for views
+      // so m_index is in the location of the slowest moving index as this is the default
+      // for RAJA...
+      template <typename... Args>
+      __snls_hdev__
+      inline
+      constexpr
+      auto&
+      operator()(Args... args) const
+      {
+         // The use of m_offset here provides us the equivalent of a rolling
+         // subview/window if our application needs it
+         return m_view(m_index, m_offset + args...);
+      }
+
+      // If we need to have like a rolling subview/window type class then
+      // we'd need some way to update the offset in our slowest moving index
+      // in the subview (so not m_view's slowest index)
+      __snls_hdev__
+      inline
+      void set_offset(const int offset) const
+      {
+         // Might want an assert in here for debugs to make sure that this is within
+         // the bounds of what m_view expects is a valid offset
+         m_offset = offset;
+      }
+
+   private:
+      // Internally we shouldn't be modifying the view itself so let's make it constant
+      const T& m_view;
+      const int m_index = 0;
+      mutable size_t m_offset = 0;
```

Comment:
Why is this mutable? We could easily change the offset setter above to non-const (usually setters are not const member functions since they modify member data), and then this wouldn't need to be mutable.

Reply:
So, I was thinking this would be more useful for when you need a constant subview that allows you to have a rolling window into your subview for specific use cases. For example, you might have a 1D subview representing state variables of multiple materials. Internally in your application, you have a loop where you iterate through the multiple materials one at a time and use the state variables for each material as if without having to have external functions have to know about the appropriate offset. We could maybe get around this by creating the get func you mentioned above, creating new subviews on the fly with the known offset. However, it seemed to me by making m_offset mutable it would make updating the subview object fairly simple if you had that const requirement for the subview due to various reasons.

Reply:
I'm still not sure I understand why this has to be mutable.  Maybe a simple test problem or example would help?  Another thought, perhaps we can remove the offset and create a separate subview (or templated on a separate subview type) that has an offset?  This makes more sense to me since it seems that most views don't really need an offset, so we're essentially storing an unused parameter here in most cases.